src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2365,27 +2369,119 @@ bool IsMethodNeededByTypeDueToPreservedScope (MethodDefinition method)
 				return false;
 
 			foreach (MethodDefinition @base in base_list) {
-				// Just because the type is marked does not mean we need interface methods.
-				// if the type is never instantiated, interfaces will be removed - but only if the optimization is enabled
-				if (@base.DeclaringType.IsInterface && Context.IsOptimizationEnabled (CodeOptimizations.UnusedInterfaces, method.DeclaringType))
+				if (!IgnoreScope (@base.DeclaringType.Scope) && !IsMethodNeededByTypeDueToPreservedScope (@base))
+					continue;
+
+				// Skip interface methods, they will be captured later by IsInterfaceMethodNeededByTypeDueToPreservedScope
+				if (@base.DeclaringType.IsInterface)
 					continue;
 
 				// If the type is marked, we need to keep overrides of abstract members defined in assemblies
-				// that are copied.  However, if the base method is virtual, then we don't need to keep the override
+				// that are copied to keep the IL valid.
+				// However, if the base method has a default implementation, then we don't need to keep the override
 				// until the type could be instantiated
 				if (!@base.IsAbstract)
 					continue;
 
-				if (IgnoreScope (@base.DeclaringType.Scope))
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Returns whether a method is needed by an interface that is in preserve scope. This should be called after interface implementations have been marked.
+		/// </summary>
+		bool IsInterfaceMethodNeededByTypeDueToPreservedScope (MethodDefinition method)
+		{
+			if (Annotations.IsMarked (method))
+				return false;
+
+			var base_list = Annotations.GetBaseMethods (method);
+			if (base_list == null)
+				return false;
+
+			foreach (MethodDefinition @base in base_list) {
+				if (!@base.DeclaringType.IsInterface)
+					continue;
+
+				if (!IgnoreScope (@base.DeclaringType.Scope)
+					&& !IsInterfaceMethodNeededByTypeDueToPreservedScope (@base)
+					&& !IsMethodNeededByTypeDueToPreservedScope (@base))
+					continue;
+				if (!Context.IsOptimizationEnabled (CodeOptimizations.UnusedInterfaces, method))
```

Comment:
Since we have different behavior for interface methods in a preserved scope depending on whether the linker is in library mode or exe mode, this was being used as a signal that we are in library mode. It is overly conservative when the linker is used in exe mode with the unused interfaces optimization off, but it is necessary for library mode to keep unmarked interface methods. Otherwise, the interface methods are seen as unmarked, so the implementation methods will be removed.
test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/StaticInterfaceMethods/StaticVirtualInterfaceMethodsInPreservedScopeLibrary.cs fails without this check.

Reply:
Ah, you're right, I was trying to make too many changes at once and something else I did caused other issues in that test.